programs/server/Server.cpp

```diff
@@ -1120,6 +1121,19 @@ int Server::main(const std::vector<std::string> & /*args*/)
             if (config->has("max_partition_size_to_drop"))
                 global_context->setMaxPartitionSizeToDrop(config->getUInt64("max_partition_size_to_drop"));
 
+            if (config->has("total_max_threads"))
+            {
+                auto total_max_threads = config->getInt("total_max_threads", 0);
+                if (total_max_threads == -1)
+                {
+                    // Based on tests total_max_threads has an optimal value when it's about 3 times of logical CPU cores
+                    constexpr size_t thread_factor = 3;
+                    total_max_threads = std::thread::hardware_concurrency() * thread_factor;
+                }
+                if (total_max_threads)
+                    ConcurrencyControl::instance().setMaxConcurrency(total_max_threads);
```

Comment:
Separate code for absent setting and zero value is required, you must pass ConcurrencyControl::Unlimited

Reply:
I think you may try in your branch set somewhere ConcurrencyControl::instance().setMaxConcurrency(50);
and you will get the similar problems.

Reply:
I pushed ConcurrencyControl::Unlimited "logic: in the latest commit.